aptos-move/framework/aptos-framework/sources/supply.move

```diff
@@ -0,0 +1,71 @@
+/// This module contains logic for upgradable and parallelizable coin supply
+/// tracking in Move.
+module aptos_framework::supply {
+    use aptos_framework::optional_aggregator::{Self, OptionalAggregator};
+    use aptos_framework::system_addresses;
+    use aptos_std::type_info;
+
+    #[test_only]
+    friend aptos_framework::supply_tests;
+
+    const MAX_U128: u128 = 340282366920938463463374607431768211455;
+
+    /// Tracks supply of coins of type `CoinType` in the system.
+    struct Supply<phantom CoinType> has store {
+        inner: OptionalAggregator,
+    }
+
+    /// Creates a new supply tracker for `CoinType`.
+    public(friend) fun new<CoinType>(): Supply<CoinType> {
+        let type_info = type_info::type_of<CoinType>();
+        let addr = type_info::account_address(&type_info);
+        new_from_address(addr)
+    }
+
+    fun new_from_address<CoinType>(addr: address): Supply<CoinType> {
+        if (system_addresses::is_aptos_framework_address(addr)) {
+            Supply {
+                // TODO: change to true once execution is working.
+                inner: optional_aggregator::new(MAX_U128, /*parallelizable=*/false),
+            }
+        } else {
+            Supply {
+                inner: optional_aggregator::new(MAX_U128, /*parallelizable=*/false),
+            }
+        }
+    }
+
+    /// Upgardes non-parallelizable supply to parallelizable. The owner of supply
+    /// is responsible for calling this function.
+    public(friend) fun upgrade<CoinType>(supply: &mut Supply<CoinType>) {
+        if (!optional_aggregator::is_parallelizable(&supply.inner)) {
+            optional_aggregator::switch(&mut supply.inner);
+        }
+    }
+
+    /// Adds `amount` to total supply of `CoinType`. Called when minting coins.
+    public(friend) fun add<CoinType>(supply: &mut Supply<CoinType>, amount: u128) {
```

Comment:
Are these functions all friend so that only coin can call these? Note that if these functions can't be easily changed to be non-friend in the future as this is considered an incompatible upgrade

Reply:
Strictly speaking, only `upgrade` must be `friend` to coin, rest can be public. I ll adapt

Reply:
Actually, @movekevin, only read should be public. If we allow public add, one can get coin info, get supply reference and pass it to some module. This module will add to supply or do something else... So we have:

Supply is only for coin, and so can be modified, created, upgraded only by coin. Read is public though.

Is this ok?

Reply:
No, even further we should not have read public, since we can call `coin::supply()`